akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -172,29 +108,16 @@ private[akka] class OutputStreamAdapter(
   private[this] def sendData(data: ByteString): Unit =
     send(() ⇒ {
       try {
-        dataQueue.put(data)
+        // FIXME take into account writeTimeout here.
+        unfulfilledDemand.acquire()
+        sendToStage(Send(data))
       } catch { case NonFatal(ex) ⇒ throw new IOException(ex) }
       if (downstreamStatus.get() == Canceled) {
```

Comment:
Shouldn't this check if the stage is cancelled be before sending? with this logic the element could potentially be accepted and sent to the stage successfully but still throw here.

What about using the async invokeWithFeedback so we get a failure on submitting instead? That could also allow us to block here until we know the element was accepted into the buffer of the stage.

Reply:
I think that would mean we could skip the atomic

Reply:
Great idea, that cleans up very nicely